### ISSUE TITLE
Improve gcomm operation in presence of high packet loss

### ISSUE THREAD

**Issue by temeo** [codership/galera#71]
This will be parent ticket for further packet loss related work.

Original report: https://bugs.launchpad.net/percona-xtradb-cluster/+bug/1274192

Introducing high delay and packet loss into network will make inter node communication highly unreliable in form of duplicated, lost or delayed messages. These kind of conditions will bring up some EVS related bugs, like currently open #37 and #40.  EVS protocol related bugs are not in the scope of this ticket and should be reported and fixed separately.

Further work can be divided roughly in three parts, which will be outlined below.

### 1. Monitoring and manual eviction

Automatic node isolation or eviction is currently not possible since EVS lacks some necessary elements like proper per-node statistics collection and protocol to communicate current view of node states without running full membership protocol. Running full membership protocol in unstable network to should be avoided as it may result in oscillating membership if the network conditions are difficult enough. Therefore the first implementation of node eviction should be based on monitoring and manual eviction until better understanding about automatic eviction has been gained.

Proposed implementation:
* Expose list of delayed node UUIDs and communication endpoints as wsrep status variables for monitoring purposes
* Implement ``evs.evict`` wsrep provider option taking list of UUIDs that should be evicted manually from cluster

### 2. Join time health check

Joining a node with bad network connection is problematic since join operation will start EVS membership protocol round which in turn performs poorly in unstable network. To avoid starting join operation over unstable network, additional health check phase for GMCast should be deviced.

When GComm protocol stack is started, joiner should first connect to all known peers in GMCast network and exchange keep alive packets to verify that the network is ok. Joining node will start upper GComm protocol layers only after health check passes. Other nodes should not treat joining node as fully qualified member in the GMCast network until joiner sends first upper level protocol packet.

### 3. Automatic eviction

After enough understanding has been gained about how to properly identify the node that is causing turbulence for group communication, automatic eviction protocol can be enabled. This work will require proper per-node statistics collection on EVS level and protocol extension to communicate statistics related view to other nodes.

**Comment by temeo** [comment]
Monitoring, manual eviction and automatic eviction from original plan has been implemented. If there is still need for join time health check, it should be reported as a separate issue.
